### PR TITLE
Implement RepeatedField.AddRange

### PR DESCRIPTION
**Pull request by jskeet**
See #1730 for context.

This PR also now optimizes AddRange when the passed collection implements ICollection.

### CODE REVIEW

**Comment by jskeet** [comment]
Reassigned to @pherl is @jtattermusch is away. The Travis build failures are the ones afflicting the whole project, not just this PR.

File: csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs [hunk 1]
```diff
@@ -75,6 +75,84 @@ public void Add_Sequence()
...
+
+        [Test]
+        public void AddRange_Optimized_NullableValueType()
+        {
+            var list = new RepeatedField<int?>();
+            list.AddRange(new List<int?> { 20, 30 });
+            Assert.AreEqual(2, list.Count);
+            Assert.AreEqual((int?) 20, list[0]);
+            Assert.AreEqual((int?) 30, list[1]);
+        }
```
Comment (anandolee) [inline csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: Add test to add range into a none empty list?

Comment (jskeet) [inline csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: Good call. Will do - but probably on Monday :)

File: csharp/src/Google.Protobuf/Collections/RepeatedField.cs [hunk 2]
```diff
@@ -300,10 +296,7 @@ public bool Remove(T item)
         /// <param name="values">The values to add to this collection.</param>
         public void Add(RepeatedField<T> values)
         {
```
Comment (anandolee) [inline csharp/src/Google.Protobuf/Collections/RepeatedField.cs]: Just wondering why not provide AddRange(RepeatedField<T> values) like IEnumerable<T> values?

File: csharp/src/Google.Protobuf/Collections/RepeatedField.cs [hunk 3]
```diff
@@ -314,20 +307,63 @@ public void Add(RepeatedField<T> values)
...
         }
 
         /// <summary>
+        /// Adds all of the specified values into this collection. This method is present to
+        /// allow repeated fields to be constructed from queries within collection initializers.
+        /// Within non-collection-initializer code, consider using the equivalent <see cref="AddRange"/>
+        /// method instead for clarity.
+        /// </summary>
+        /// <param name="values">The values to add to this collection.</param>
+        public void Add(IEnumerable<T> values)
```
Comment (anandolee) [inline csharp/src/Google.Protobuf/Collections/RepeatedField.cs]: I'd prefer to only provide one instead of both unless it needs users' migration?

File: csharp/src/Google.Protobuf/Collections/RepeatedField.cs (hunk 2 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/RepeatedField.cs]: Yes, could do that. Or we could remove the overload for Add, and just detect it as another kind of  optimization in AddRange. Very happy to do that.

File: csharp/src/Google.Protobuf/Collections/RepeatedField.cs (hunk 3 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/RepeatedField.cs]: No, it's very deliberate that there are both. In idiomatic code, `AddRange` is what you'd call manually, but `Add` would be called implicitly by the compiler:

```csharp
// Collection initializer
var person = new Person { Friends = { other.Friends.Where(f => f.Name == "Fred") } };
// Manual code, if you're modifying the message after construction
person.Friends.AddRange(/* something other collection */);
```

File: csharp/src/Google.Protobuf/Collections/RepeatedField.cs (hunk 2 above)
Comment (jskeet) [inline csharp/src/Google.Protobuf/Collections/RepeatedField.cs]: Removed the overload for Add, and put the code in AddRange instead. Going through CI now...

**Comment by jskeet** [comment]
@chrisdunelm - could you review this, as @jtattermusch is unavailable at the moment? I can walk you through any bits you're unfamiliar with.

File: csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs [hunk 4]
```diff
@@ -75,10 +75,96 @@ public void Add_Sequence()
         }
 
         [Test]
-        public void Add_RepeatedField()
+        public void AddRange_SlowPath()
+        {
+            var list = new RepeatedField<string>();
+            list.AddRange(new[] { "foo", "bar" }.Select(x => x));
```
Comment (chrisdunelm) [inline csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: Arrays implement ICollection, so this isn't testing the slow-path/

Comment (jskeet) [inline csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: That's why I'm calling `Select` :)

Comment (chrisdunelm) [inline csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: [Hides from shame, must read code better]

**Comment by chrisdunelm** [comment]
LGTM
